File: cc/maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.cc
#include <bits/stdc++.h>
using namespace std;
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
typedef TreeNode *node;
class Solution {
public:
  void dfs(node root, int &dep) {
    if(!root) return;
    if(!root->left && !root->right) {
      dep = max(dep, root->val);
      return;
    }
    if(root->left) {
      root->left->val = root->val + 1;
      dfs(root->left, dep);
    }
    if(root->right) {
      root->right->val = root->val + 1;
      dfs(root->right, dep);
    }
  }

  /*
  int maxDepth(node root) {
    if(!root) return 0;
    root->val = 1;
    int res = 1;
    dfs(root, res);
    return res;
  }
  */

  // elegent
  int maxDepth(node root) {
    return root ? 1+max(maxDepth(root->left), maxDepth(root->right)):0;
  }
};

//const int N = ((1<<5)-1);
const int N = 15;
int main(int argc, const char *argv[])
{
  int A[N+1] = 
  { 0,
    5,
    4,8,
    11,0,13,4,
    7,2,0,0,0,0,5,1
  };
  //A[1] = 1;
  //A[2] = 2;
  //A[4] = 3;
  //A[8] = 4;
  //A[16] = 5;
  node nodes[N+1] = { 0 };
  for(int i=1; i<N+1; i++) {
    if(A[i]) nodes[i] = new TreeNode(A[i]);
  }
  for(int i=1; i<N+1; i++) {
    if(2*i < N+1 && nodes[i]) nodes[i]->left = nodes[2*i];
    if(2*i+1 < N+1 && nodes[i]) nodes[i]->right = nodes[2*i+1];
  }
  puts("level order");
  int lv = 0;
  while((1<<lv) <= N) ++lv;
  for(int i=0; i<lv; i++) {
    for(int j=0; j<(1<<i) && (1<<i)+j < N+1; j++) 
      if(nodes[(1<<i)+j]) printf("%d ", nodes[(1<<i)+j]->val);
      else printf("# ");
    puts("");
  }
  puts("--------");

  Solution so;
  cout << so.maxDepth(nodes[1]) << endl;
  return 0;
}

File: rs/src/quests/maximum_depth_of_binary_tree.rs
/**
 * [104] Maximum Depth of Binary Tree
 *
 * Given the root of a binary tree, return its maximum depth.
A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
 
Example 1:
Input: root = [3,9,20,null,null,15,7]
Output: 3

Example 2:
Input: root = [1,null,2]
Output: 2

 
Constraints:

	The number of nodes in the tree is in the range [0, 104].
	-100 <= Node.val <= 100

 */
pub struct Solution {}
use super::super::util::tree::{TreeNode, to_tree};

// submission codes start here

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::cmp::max;
impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        root.map(|node| {
            let left = Self::max_depth(node.borrow_mut().left.take());
            let right = Self::max_depth(node.borrow_mut().right.take());
            1 + max(left, right)
        }).unwrap_or(0)
    }
}

// submission codes end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        assert_eq!(2, Solution::max_depth(tree![1,null,2]));
    }
}
